File: september2019/twoSum/sol.c
#include <stdlib.h>
#include <limits.h>
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* twoSum(int* nums, int numsSize, int target, int* returnSize){
  *(returnSize) = 2;
  int maxNum = INT_MAX;
  int* sol = malloc(sizeof(int) * *(returnSize));
  *(sol) = 4;
  int* hash = calloc(sizeof(int), maxNum*2);
  for (int i = 0; i < numsSize; i++) {
    if (hash[target - *(nums + i) + maxNum] ) {
      *(sol) = hash[target - *(nums + i) + maxNum] - 1;
      *(sol + 1) = i;
      return sol;
    } else
      hash[*(nums + i) + maxNum] = i + 1;
  }
  // To make the compiler happy.
  return sol;
}

File: april2020/Product_of_Array_Except_Self/sol.go
package main

import (
	"fmt"
)

func productExceptSelf(nums []int) []int {
	r := make([]int, len(nums))
	l := make([]int, len(nums))
	rp := 1
	lp := 1

	for i := 0; i < len(nums); i++ {
		r[i] = rp * nums[i]
		rp = r[i]
		l[len(nums) - 1 - i] = lp * nums[len(nums) - 1 - i]
		lp = l[len(nums) - 1 - i]
	}

	sol := make([]int, len(nums))

	for i := 0; i < len(nums); i++ {
		if i > 0 && i < len(nums) - 1 {
			sol[i] = r[i - 1]*l[i + 1] 
			continue
		} 
		if i > 0 {
			sol[i] = r[i - 1]
		}
		if i < len(nums) - 1 {
			sol[i] = l[i + 1]
		}
	}

	return sol
}

func main() {
	fmt.Println(productExceptSelf([]int{1,2}))
}